### PR TITLE
Boot Manager requirements

### PR DESCRIPTION
**Pull request by vstehle**
Let's clarify the requirements on the UEFI Boot Manager to address #130.

### CODE REVIEW

**Comment by vstehle** [comment]
As per [EBBR call of Aug 26](https://github.com/ARM-software/ebbr/wiki/EBBR-Notes-2024.08.26) this must be respun with better wording for `BootOptionsSupport`.

**Comment by vstehle** [comment]
Changes:
- Tighten wording around `Boot*` variables, as per feedbacks during the [EBBR call of Aug 26](https://github.com/ARM-software/ebbr/wiki/EBBR-Notes-2024.08.26).
- Mention the optional `Driver####`, `SysPrep####`, `OsRecovery####` and `PlatformRecovery####` variables explicitly, to help searching the UEFI spec.
- Warn that `BootOptionSupport` will be required in the future.

File: source/chapter2-uefi.rst [hunk 1]
```diff
@@ -178,6 +178,8 @@ interface specific UEFI protocols, and so they have been made optional.
        For this reason EBBR implementations are not required to support option
        ROM loading.
 
+.. _section-required-global-vars:
```
Comment (xypron) [inline source/chapter2-uefi.rst:181]: This seems to be an unrelated change that should be in a separate patch.

File: source/chapter2-uefi.rst [hunk 2]
```diff
@@ -666,3 +668,71 @@ the `GetNextHighMonotonicCount()` runtime service. [#BootNote]_
...
+   * - Required element
+     - Description
+   * - Load options
+     - The load options residing in the `Boot####` variables, the priorities set
+       by the `BootNext` and `BootOrder` variables and the reporting of the
+       selected boot option with the `BootCurrent` variable, as defined in
+       :UEFI:`3.1`, must be supported (see :ref:`section-required-global-vars`).
+   * - Default behavior
+     - The default boot behavior after trying `BootOrder` as defined in
+       :UEFI:`3.4.3` must be supported.
```
Comment (xypron) [inline source/chapter2-uefi.rst:696]: Please, provide the section title.

File: source/chapter2-uefi.rst [hunk 3]
```diff
@@ -666,3 +668,71 @@ the `GetNextHighMonotonicCount()` runtime service. [#BootNote]_
...
+   * - Boot mechanisms
+     - Booting an application from device as defined in :UEFI:`3.5` must be
```
Comment (xypron) [inline source/chapter2-uefi.rst:698]: Please, provide the section title as numbering tends to change between UEFI spec revisions.

File: source/chapter2-uefi.rst [hunk 4]
```diff
@@ -666,3 +668,71 @@ the `GetNextHighMonotonicCount()` runtime service. [#BootNote]_
...
+
+   * - Required element
+     - Description
+   * - Load options
+     - The load options residing in the `Boot####` variables, the priorities set
+       by the `BootNext` and `BootOrder` variables and the reporting of the
+       selected boot option with the `BootCurrent` variable, as defined in
+       :UEFI:`3.1`, must be supported (see :ref:`section-required-global-vars`).
+   * - Default behavior
+     - The default boot behavior after trying `BootOrder` as defined in
```
[the remark this answers is not in the document]
Comment (xypron) [reply source/chapter2-uefi.rst:695]: What from the section 3.4.3. "Boot Option Variables Default Boot Behavior" are you actually requiring? Below you seem to be excluding everything mentioned in the section.

File: source/chapter2-uefi.rst [hunk 5]
```diff
@@ -666,3 +668,71 @@ the `GetNextHighMonotonicCount()` runtime service. [#BootNote]_
...
+       not required.
+   * - System preparation applications
+     - Launching system preparation applications and the associated
+       `SysPrep####` variables as defined in :UEFI:`3.1.7` are not required.
+   * - `EFI_BOOT_MANAGER_POLICY_PROTOCOL`
+     - This protocol defined in :UEFI:`3.2` is not required.
+   * - Boot Option Recovery
+     - The OS-defined recovery and the associated `OsRecovery####` variables as
+       defined in :UEFI:`3.4.1`, and the platform-specific recovery and
+       associated `PlatformRecovery####` variables as defined in :UEFI:`3.4.2`
```
Comment (xypron) [inline source/chapter2-uefi.rst:734]: What about 3.4.3. "Boot Option Variables Default Boot Behavior" requirements which also refer to PlatformRecovery####. Requiring 3.4.3 behavior above and excluding 3.4.2 behavior confuses me.

File: source/chapter2-uefi.rst (hunk 2 above)
Comment (vstehle) [reply source/chapter2-uefi.rst:696]: Addressed during [the call of Sep 9](https://github.com/ARM-software/ebbr/wiki/EBBR-Notes-2024.09.09); the [Sphinx extension](https://github.com/ARM-software/ebbr?tab=readme-ov-file#sphinx-extensions) does this automatically.

File: source/chapter2-uefi.rst (hunk 3 above)
Comment (vstehle) [reply source/chapter2-uefi.rst:698]: Addressed during the [call of Sep 9](https://github.com/ARM-software/ebbr/wiki/EBBR-Notes-2024.09.09); the [Sphinx extension](https://github.com/ARM-software/ebbr?tab=readme-ov-file#sphinx-extensions) adds the title automatically.

File: source/chapter2-uefi.rst (hunk 1 above)
Comment (vstehle) [reply source/chapter2-uefi.rst:181]: Addressed during the [call of Sep 9](https://github.com/ARM-software/ebbr/wiki/EBBR-Notes-2024.09.09); we add the sections anchors only as necessary.

File: source/chapter2-uefi.rst (hunk 5 above)
Comment (vstehle) [reply source/chapter2-uefi.rst:734]: Fixed, thanks.

File: source/chapter2-uefi.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (vstehle) [reply source/chapter2-uefi.rst:695]: I have reworked this part; please let me know how you like it now.

**Comment by vstehle** [comment]
Reworked after @xypron's comments (thanks!) and [our call of Oct. 7](https://github.com/ARM-software/ebbr/wiki/EBBR-Notes-2024.10.07).
Please review and let me know how you like it.

**Comment by vstehle** [comment]
Rebased on CI fixes.

**Comment by xypron** [comment]
@vstehle 

Concerning BootOptionSupport the UEFI spec has "If the global variable is not present, then an installer or application must act as if a value of 0 was returned."

What relevance has your warning about requiring BootOptionSupport in this context?

**Comment by vstehle** [comment]
> What relevance has your warning about requiring BootOptionSupport in this context?

Hi @xypron, I think you are right; all the wording around `BootOptionSupport` is confusing. As this variable is optional in UEFI already, let's simply leave it as it is.
While at it, we can do the same for all the other boot manager features, already optional in UEFI.
I think this makes the addition lighter; let me know how you like it.

**Comment by vstehle** [comment]
Changes:
- Removed the two warnings about `BootOptionSupport` becoming mandatory in the future.
- List only the features we are making optional in the second table. Therefore removed the following mentions:
  - Boot menu
  - Load option categories
  - `BootOptionSupport`
  - `Key####` hot keys
  - `SysPrep*` system preparation applications
